src/Dependencies/Collections/ImmutableSegmentedDictionary`2+ValueBuilder.cs

```diff
@@ -0,0 +1,244 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    internal readonly partial struct ImmutableSegmentedDictionary<TKey, TValue>
+    {
+        private struct ValueBuilder : IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>, IDictionary
+        {
+            /// <summary>
+            /// The immutable collection this builder is based on.
+            /// </summary>
+            private ImmutableSegmentedDictionary<TKey, TValue> _dictionary;
+
+            /// <summary>
+            /// The current mutable collection this builder is operating on. This field is initialized to a copy of
+            /// <see cref="_dictionary"/> the first time a change is made.
+            /// </summary>
+            private SegmentedDictionary<TKey, TValue>? _mutableDictionary;
+
+            internal ValueBuilder(ImmutableSegmentedDictionary<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+                _mutableDictionary = null;
+            }
+
+            public IEqualityComparer<TKey> KeyComparer
+            {
+                get
+                {
+                    return ReadOnlyDictionary.Comparer;
+                }
+
+                set
+                {
+                    if (value is null)
+                        throw new ArgumentNullException(nameof(value));
+
+                    if (value != KeyComparer)
+                    {
+                        // Rewrite the mutable dictionary using a new comparer
+                        var valuesToAdd = ReadOnlyDictionary;
+                        _mutableDictionary = new SegmentedDictionary<TKey, TValue>(value);
+                        AddRange(valuesToAdd);
+                    }
+                }
+            }
+
+            public int Count => ReadOnlyDictionary.Count;
+
+            internal SegmentedDictionary<TKey, TValue> ReadOnlyDictionary => _mutableDictionary ?? _dictionary._dictionary;
+
+            IEnumerable<TKey> IReadOnlyDictionary<TKey, TValue>.Keys => throw new NotSupportedException();
```

Comment:
📝 It would be preferable for `Keys` and `Values` to work, but I couldn't think of a clean way for that to happen